### PR TITLE
[MST-1473] Proctoring Service Launch Implementation

### PR DESCRIPTION
**Pull request by MichaelRoytman**
### Description

This pull request implements the [LTI Proctoring Services](http://www.imsglobal.org/spec/proctoring/v1p0) standard LTI messages and proctoring flow. It does not implement the Assessment Control Service.

This pull request relies on the `location` being fully decoupled from the XBlock (https://github.com/openedx/xblock-lti-consumer/issues/273) and on the ability for consuming contexts to pass relevant context to the the library. As such, this is a work-in-progress and will evolve as those requirements are met.

Because of the dependency on the XBlock, you will find a number of TODOs in this pull request. I am hoping to address these iteratively as I receive feedback.

I have also not added tests yet, to avoid churn.

### Key Details

* I added a set of views that implement the LTI Proctoring Services launch flow: `start_proctoring`, `launch_gate_endpoint_proctoring`, `start_assessment` and `end_assessment`. You'll note that `launch_gate_endpoint_proctoring` has a lot of overlap with `launch_gate_endpoint` in https://github.com/openedx/xblock-lti-consumer/pull/254. Because of the dependency on the XBlock, I could not easily reuse the logic there; I felt having a separate view for now would be easier.
* I added an `LtiProctoringConsumer`.
* I added the above views to the URL configuration.

### Dependencies
* https://github.com/openedx/xblock-lti-consumer/pull/254

### Testing Instructions

**Setup Steps:**

1. Clone the [xblock-lti-consumer](https://github.com/openedx/xblock-lti-consumer) repository into your `{workspace}/src` directory. `{workspace}` should be the directory in which you have your other edX git repositories and should be the same directory your `devstack` repository is located in.
2. Check out this branch (`mroytman/MST-1473-proctoring-service-launch`) locally.
3. Clone the [edx-exams](https://github.com/edx/edx-exams) repository into your `{workspace}` directory. `{workspace}` should be the directory in which you have your other edX git repositories and should be the same directory your `devstack` repository is located in.
4. Check out the branch `mroytman/MST-1473-proctoring-service-launch` locally.
5. Follow the [One Time Setup](https://github.com/edx/edx-exams#one-time-setup) instructions.

**Testing Instructions:**

If you want to test this implementation against the IMS Proctoring Specification testing tool, follow these steps.

1. Sign up for an IMS account [here](https://site.imsglobal.org/user/register), if you are not already a member. The approval process should take around a day.
2. You will need to set up port forwarding so that your running Django server can connect with the IMS testing tool. You can use `ngrok` for this or any other tool you like. `ngrok` has comprehensive documentation for how to accomplish this.
3. In your local checkout of `mroytman/MST-1473-proctoring-service-launch` in `edx-exams`, change `LMS_ROOT_URL` in `local.py` to the forwarded URL. For example, `https://9c11-73-238-153-173.ngrok.io`.
4. Once your membership has been approved, go to the testing tool [here](https://certification.imsglobal.org/proctoring/launch).
5. In the testing tool, click `Get Started`.
6. Enter your name, email, and organization and click `Confirm`.
7.  Click `Create` under `Create New Product`.
8. Enter a `Product` and `Version` and click `Confirm`. These values can be anything for the purposes of testing.
9. Click `Select` under `Assessment Platform`.
11. Click `Begin Configuration`.
12. Open up the Django admin by visiting `localhost:18740/admin` and logging in with the default edX development login credentials.
13. Under `LTI_CONSUMER > Lti configuration`, click `Add`.
14. Enter the following configuration:
15.1. `Version`: `LTI 1.3 (with LTI Advantage Support)`
13.2. `Config store`: `Configuration Stored on this model`.
16. From the testing tool, copy the following configuration into the `Lti configuration` model in the Django admin.
14.1. `Tool End-Point to Receive Launches` > `Lti 1p3 launch url`
14.2. `OIDC Login Initiation URL` > `Lti 1p3 oidc url`
17. Click `Next`.
18. Enter the following configuration in the testing tool. In this example, `{root}` refers to the URL that is being forwarded. For example, `https://9c11-73-238-153-173.ngrok.io`.
16.1. `Testing iss Value`: `{root}`
16.2. `OIDC Auth URL`: `{root}/lti/launch_proctoring`
16.3. `Platform Well-Known/JWKS URL`: `{root}/lti/public_keyset`
16.4. `Client Id (Assigned to Cert Suite)` > Enter the `Lti 1p3 client id` from the `Lti configuration` model in the Django admin.
19. Click `Next`.
20. Enter `1` for the deployment ID and click `Next`.
21. From the testing tool, copy the`Proctoring Tool JWKS URL` into the `Lti configuration` model in the Django admin in the `Lti 1p3 tool keyset url` field.
22. Click `Next`.
23. Check `Pause on OIDC Exchanges to Inspect Values` and/or `Pause Prior to JWT Messages to Inspect Values` depending on whether you want interstitials before Tool requests or responses to the Assessment Platform.
24. Visit `{root}/api/start_proctoring` to kick off the flow.

### CODE REVIEW

File: lti_consumer/models.py [hunk 1]
```diff
@@ -235,8 +242,10 @@ def clean(self):
             raise ValidationError({
                 "config_store": _("LTI Configuration stores on XBlock needs a block location set."),
             })
+        # If we're in an XBlock context, then check whether the database_config_enabled flag is enabled. Otherwise,
+        # don't restrict saving LtiConfiguration.
         if self.version == self.LTI_1P3 and self.config_store == self.CONFIG_ON_DB:
-            if not database_config_enabled(self.block.location.course_key):
+            if self.location and not database_config_enabled(self.block.location.course_key):
```
Comment (MichaelRoytman) [inline lti_consumer/models.py:248]: I'm going to pull this into a separate pull request, but I needed it here to get it working.

File: lti_consumer/plugin/views.py [hunk 2]
```diff
@@ -501,3 +556,430 @@ def list(self, *args, **kwargs):
...
+    # TODO: This will need to have additional consideration for PII.
+    # optional_user_identity_claims = get_optional_user_identity_claims()
+    # lti_consumer.set_proctoring_user_data(
+    #     **optional_user_identity_claims
+    # )
+
+    try:
+        lti_response = lti_consumer.check_and_decode_token(token)
+    # TODO: This needs better error handling.
+    except (BadJwtSignature, MalformedJwtToken, NoSuitableKeys):
```
Comment (ashultz0) [inline lti_consumer/plugin/views.py:768]: a nonspecific 400 is an ok return for bad signature / malformed jwt both of which could be attacks

File: lti_consumer/plugin/views.py [hunk 3]
```diff
@@ -501,3 +556,430 @@ def list(self, *args, **kwargs):
...
+        end_assessment_return_key = get_cache_key(app="lti", key="end_assessment_return", user_id=request.user.id)
+        # We convert the boolean to an int because memcached will return an int even if a boolean is stored. This
+        # ensures a consistent return value.
+        end_assessment_return_value = int(end_assessment_return)
+        TieredCache.set_all_tiers(end_assessment_return_key, end_assessment_return_value)
+
+    return JsonResponse(data={})
+
+
+# We do not want Django's CSRF protection enabled for POSTs made by external services to this endpoint. This is because
```
[the remark this answers is not in the document]
Comment (ashultz0) [reply lti_consumer/plugin/views.py:788]: I don't think that django's CSRF not being able to handle this case is worth an ADR. It might be worth separating this note out into some readme if we have a suitable one (but I don't think we do)

I honestly would have just written "django's CSRF can't handle a flow where it isn't in control of the middle" so this is already more info than I demand :)

File: lti_consumer/plugin/views.py [hunk 4]
```diff
@@ -501,3 +556,430 @@ def list(self, *args, **kwargs):
...
+# https://openid.net/specs/openid-connect-core-1_0.html#AuthRequest
+@require_http_methods(["GET", "POST"])
+def launch_gate_endpoint_proctoring(request, suffix=None):  # pylint: disable=unused-argument
+    """
+    This is the Assessment Platform's OIDC login authentication/authorization endpoint.
+
+    This uses the "client_id" query parameter or form data to identify the LtiConfiguration and its consumer to generate
+    the LTI 1.3 Launch Form.
+    """
+    preflight_response = request.GET.dict() if request.method == 'GET' else request.POST.dict()
```
Comment (ashultz0) [inline lti_consumer/plugin/views.py:822]: this is super weird, there isn't a request method that just does this? request.THING_YOU_HAVE.dict()?

File: lti_consumer/plugin/views.py [hunk 5]
```diff
@@ -501,3 +556,430 @@ def list(self, *args, **kwargs):
...
+
+        return render(request, 'html/lti_1p3_launch.html', context)
+    except Lti1p3Exception as exc:
+        log.warning(
+            "Error preparing LTI 1.3 launch for client_id %s: %s",
+            client_id,
+            exc,
+            exc_info=True
+        )
+        return render(request, 'html/lti_1p3_launch_error.html', context, status=HTTP_400_BAD_REQUEST)
```
Comment (ashultz0) [inline lti_consumer/plugin/views.py:985]: this function is pretty big and could do with some chunking, maybe pull the try: contents out into its own thing

File: lti_consumer/plugin/views.py [hunk 6]
```diff
@@ -501,3 +512,398 @@ def list(self, *args, **kwargs):
                 "error": "above_response_limit",
                 "explanation": "The number of retrieved users is bigger than the maximum allowed in the configuration.",
             }, status=HTTP_403_FORBIDDEN)
+
```
[the remark this answers is not in the document]
Comment (ashultz0) [reply lti_consumer/plugin/views.py:679]: the import renaming in exams is a bit of an indicator of weird stuff:
from lti_consumer.plugin.views import start_proctoring as lti_start_proctoring

would we expect any app to ever be able to use these views without wrapping? If not can we get any benefit by having them not be structured as views? They seem to all do request param copying and the exam wrapper also does that, so we make a copy of a copy to add things in layers. Maybe these functions could just take a dictionary of extra stuff to add. Maybe that could go all the way down into proctoring_preflight because all of its callers do the same setup

File: lti_consumer/lti_1p3/consumer.py [hunk 7]
```diff
@@ -659,3 +661,228 @@ def enable_nrps(self, context_memberships_url):
...
+            tool_key,
+            tool_keyset_url
+        )
+        self.proctoring_data = {}
+
+    def set_proctoring_data(self, **kwargs):
+        """
+        Set the self.proctoring_data dictionary with the provided kwargs, so long as a given key is in
+        LTI_PROCTORING_DATA_KEYS.
+        """
```
Comment (ashultz0) [inline lti_consumer/lti_1p3/consumer.py:712]: If the kwargs are an incomplete set, what happens? Should this be more strongly typed with actual args?

File: lti_consumer/lti_1p3/consumer.py [hunk 8]
```diff
@@ -659,3 +661,228 @@ def enable_nrps(self, context_memberships_url):
...
+        # If the picture attribute is provided it MUST point to a picture taken by the Proctoring Tool.
+        # It MUST NOT be the same picture provided by the Assessment Platform in the Start Proctoring message.
+        picture = verified_user.get('picture')
+        if picture and picture == self.lti_claim_user_data.get('picture'):
+            raise exceptions.InvalidClaimValue(
+                'If the verified_claim is provided and contains the picture claim,'
+                ' the picture claim should not be the same picture provided by the Assessment Platform to the Tool.'
+            )
+        # TODO: We can leverage these verified user claims. For example, we could use
+        #       these claims for Name Affirmation.
```
Comment (ashultz0) [inline lti_consumer/lti_1p3/consumer.py:877]: it might make sense to pull out the claim check in to a separate function which would be easier to test by itself

it could return the verified user we could just get that out again

File: lti_consumer/plugin/views.py [hunk 9]
```diff
@@ -501,3 +512,398 @@ def list(self, *args, **kwargs):
...
+    #       lti_consumer.set_user_data uses the regular roles. So, for now, I'm leaving it empty.
+    # Required user claim data
+    lti_consumer.set_user_data(
+        user_id=request.user.id,
+        role=None,
+    )
+
+    # These claims are optional. They are necessary to set in order to properly verify the verified_user claim,
+    # if the Proctoring Tool includes it in the JWT.
+    # TODO: This will need to have additional consideration for PII.
```
[the remark this answers is not in the document]
Comment (Zacharis278) [reply lti_consumer/plugin/views.py:879]: I don't think there's an issue here because we aren't allowing course teams to configure their own providers. Any available vendor will have a privacy agreement with us.

File: lti_consumer/api.py [hunk 10]
```diff
@@ -142,25 +142,14 @@ def get_lti_1p3_launch_info(config_id=None, block=None):
     }
 
 
-def get_lti_1p3_launch_start_url(config_id=None, block=None, deep_link_launch=False, dl_content_id=None, hint=""):
+def get_lti_1p3_launch_start_url(config_id=None, block=None, lti_hint="", hint=""):
```
[the remark this answers is not in the document]
Comment (Zacharis278) [reply lti_consumer/api.py:147]: would it be appropriate to change `hint` to `login_hint`? I think that makes the distinction between the parameters clearer

File: lti_consumer/plugin/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zacharis278) [reply lti_consumer/plugin/views.py:679]: Do we absolutely need edx-exams to be providing the `attempt_number` and such? Could the UI just provide these values when starting the flow? In that way we can leave the views entirely within the lti-consumer library and avoid wrapping or using api functions right?

File: lti_consumer/plugin/views.py [hunk 11]
```diff
@@ -501,3 +512,398 @@ def list(self, *args, **kwargs):
...
+    #           Proctoring Tool to identify the LtiConfiguration; iss should match the client_id of the
+    #           LtiConfiguration. Although there is a risk that the JWT is not authentic, we will validate the
+    #           authenticity of the JWT and raise an exception if the signature is invalid later in this function.
+    token = request.POST.get('JWT')
+    jwt = JWT().unpack(token)
+
+    client_id = jwt.payload().get('iss')
+
+    try:
+        lti_config = LtiConfiguration.objects.get(lti_1p3_client_id=client_id)
```
[the remark this answers is not in the document]
Comment (Zacharis278) [reply lti_consumer/plugin/views.py:825]: I think _technically_ this has to be paired with the deployment id to be guaranteed unique. But it also looks like we don't support multiple deployments of a tool so maybe just a comment acknowledging that. Or we hard code this as deployment_id="1" to future proof ourselves.

File: lti_consumer/plugin/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MichaelRoytman) [reply lti_consumer/plugin/views.py:679]: What would that look like? The `launch_gate_endpoint_proctoring` view should be invoked when the Tool makes a third party authentication request to the Platform (i.e. `edx-exams`). How would the frontend pass these data into the flow? I could see how the frontend could pass `attempt_number` when it kicks off the LTI flow (via the `start_proctoring` view) when, say, a learner presses a "start" button, but that data isn't needed the LTI flow is being kicked off. Once the flow is kicked off, will the frontend even be involved until it comes time to render content?

[the remark this answers is not in the document]
Comment (MichaelRoytman) [reply lti_consumer/plugin/views.py:679]: I'm going to call "LTI views" the endpoints from `xblock-lti-consumer` that are installed into `edx-exams`.

For the first request that is made, which is the third party login initiation (i.e. `start_proctoring`), the frontend _could_ make a request to the LTI view that would include the LTI `config_id`. That makes sense to me. I assume the frontend would get it in some backend-for-frontend request before rendering the exam, right?

The Platform receives this request, grabs the `LtiConfiguration` using the `config_id`, crafts a preflight URL using the `LtiProctoringConsumer`, and returns a redirect to that preflight URL. The browser makes the request. Next, the Tool gets the request and makes a third party authentication request to the LTI view through the browser. `edx-exams` gets the request and has to issue an LTI launch. At this point, the LTI view needs to know `attempt_number`, `resource_link`, etc. in order to issue a launch. These data come from `edx-exams`, but the LTI view has no knowledge of them. I don't think the frontend can help here, unless I'm mistaken?

File: lti_consumer/api.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (MichaelRoytman) [reply lti_consumer/api.py:147]: I definitely support this. I found the distinction confusing myself and had to keep checking which was which. I'll just echo the names as they are in the specification: `login_hint` and `lti_message_hint`.

File: lti_consumer/lti_1p3/consumer.py (hunk 8 above)
Comment (MichaelRoytman) [reply lti_consumer/lti_1p3/consumer.py:877]: Are you referring to lines 856 - 877?

File: lti_consumer/plugin/views.py (hunk 4 above)
Comment (MichaelRoytman) [reply lti_consumer/plugin/views.py:942]: I couldn't find one, no ☹️ .

File: lti_consumer/lti_1p3/consumer.py (hunk 8 above)
Comment (ashultz0) [reply lti_consumer/lti_1p3/consumer.py:877]: yep!

File: lti_consumer/plugin/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (MichaelRoytman) [reply lti_consumer/plugin/views.py:679]: @ashultz0 

> would we expect any app to ever be able to use these views without wrapping? If not can we get any benefit by having them not be structured as views? They seem to all do request param copying and the exam wrapper also does that, so we make a copy of a copy to add things in layers. Maybe these functions could just take a dictionary of extra stuff to add. Maybe that could go all the way down into proctoring_preflight because all of its callers do the same setup

Unless we can find an alternative way to pass the necessary data around, I don't think so. I can see what that your suggestion would look like. You're envisioning that the input to any of these function would be a dictionary (not a request), and then the output would be a response object, right?

File: lti_consumer/lti_1p3/consumer.py [hunk 12]
```diff
@@ -659,3 +661,228 @@ def enable_nrps(self, context_memberships_url):
...
+        This method validates the JWT signature and decodes the JWT. It also validates the claims in the JWT according
+        to the Proctoring Services specification.
+
+        It either returns a dictionary containing information required by the start_assessment, or it raises an
+        exception if any claims is missing or invalid.
+        """
+        # Decode token and check expiration.
+        proctoring_response = self.tool_jwt.validate_and_decode(token)
+
+        # TODO: We MUST perform other forms of validation here.
```
Comment (MichaelRoytman) [inline lti_consumer/lti_1p3/consumer.py:794]: I reread the specification, and now I cannot find where I read that we **MUST** do this validation. I swore I read it, but I cannot find it again. Based on that, I think the majority of this validation is not necessary. I may have misread a paragraph like the following as a requirement that we validate that the tool did its job. I don't think we _need_ to do that. That would mean that the `attempt_number` and `resource_link` would not longer be needed by the `start_assessment` view.  

> 4.3.1.6 Attempt number claim
The https://purl.imsglobal.org/spec/lti-ap/claim/attempt_number claim specifies the candidate's attempt number, as set by the Assessment Platform.  The Proctoring Tool MUST copy this value to the Start Assessment message unchanged.

File: lti_consumer/plugin/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ashultz0) [reply lti_consumer/plugin/views.py:679]: I think I meant instead of each layer passing in a request and a dict and combining the dict into the request params to make new request params we would just pass the request and ever increasing set of new params down to the lowest level before sticking them together and getting the response. The various layers all have the view interface but... is anyone ever really going to use the inner views as views? If not, they don't have the view argument set.

File: lti_consumer/lti_1p3/consumer.py (hunk 7 above)
Comment (MichaelRoytman) [reply lti_consumer/lti_1p3/consumer.py:712]: I think I did it this way originally because I thought we'd need partial data in some places, but that's no longer the case. I'll change it back to something like `set_proctoring_data(attempt_number, resource_link, ...)`.

File: lti_consumer/plugin/views.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zacharis278) [reply lti_consumer/plugin/views.py:679]: > I'm going to call "LTI views" the endpoints from `xblock-lti-consumer` that are installed into `edx-exams`.
> 
> For the first request that is made, which is the third party login initiation (i.e. `start_proctoring`), the frontend _could_ make a request to the LTI view that would include the LTI `config_id`. That makes sense to me. I assume the frontend would get it in some backend-for-frontend request before rendering the exam, right?
> 
> The Platform receives this request, grabs the `LtiConfiguration` using the `config_id`, crafts a preflight URL using the `LtiProctoringConsumer`, and returns a redirect to that preflight URL. The browser makes the request. Next, the Tool gets the request and makes a third party authentication request to the LTI view through the browser. `edx-exams` gets the request and has to issue an LTI launch. At this point, the LTI view needs to know `attempt_number`, `resource_link`, etc. in order to issue a launch. These data come from `edx-exams`, but the LTI view has no knowledge of them. I don't think the frontend can help here, unless I'm mistaken?

I guess I was under the impression that the contextual details such as attempt and resource_id wood be somehow be passed through the initial authentication steps. Otherwise once we do get to the step where the tool requests a launch from edx-exams how would edx exams even know what specific attempt this is for if it's only context at that point is a certain LtiConfiguration?

[the remark this answers is not in the document]
Comment (MichaelRoytman) [reply lti_consumer/plugin/views.py:679]: They're not, unfortunately.

The specification doesn't detail how to accomplish this. It's a good question. I agree that it's not really clear how to get much more than the currently requesting user and the provider being used for the exam (from the `client_id`) when crafting the LTI launch message.

I believe that the way the existing LTI 1.3 launch gets around this is by being coupled to the XBlock `location` and by misusing the `login_hint` parameter. The `login_hint` parameter refers to a `location` on the `LtiConfiguration`, but it also lets the library pull data about the context of the LTI launch through the XBlock. I think that's why the full decoupling of the LTI views from `LtiConfiguration.location` has been challenging. It's a little bit like apples and oranges, but the `context_claim` is similar to `attempt_number` or `resource_link`. I considered whether we could abuse the `login_hint` to refer to an attempt to tie the two requests together, but it's really not the purpose of the parameter. Nevertheless, we _could_ do that, but it doesn't address the other data that the library will need (e.g. `resource_link`, user `role`, etc.).

I'll think through some options tomorrow and report back.

**Comment by mphilbrick211** [comment]
@MichaelRoytman just checking in on this - are you planning to proceed with this PR? If so, you will need to re-run the tests. Thank you!

**Comment by MichaelRoytman** [comment]
@mphilbrick211 Thanks for following up. I do not plan to merge this pull request. I'll close it out. I have another [pull request](https://github.com/openedx/xblock-lti-consumer/pull/297) open that will take this pull request's place, and I have put it in review now. I'm hoping to merge it in soon.
